### DISCUSSION TITLE
SetWarmStart function

### DISCUSSION THREAD

**Discussion by ycpengpeng** [coin-or/Cbc#568] [Discussion]
Hi, I want to set an initial solution of the integer linear programming problem because I want the final solution to be near to the initial solution. I use the OsiClpSolverInterface model and the SetWarmStart function. My code is as follows:
```c++
OsiClpSolverInteface model;
model.loadProblem(numcols, numrows, start, index values, collb,NULL,obj,rowlb,rowub);
model.setInteger(0);
model.setObjSense(1.0);
CbcModel solver(model);
// variable_size is the number of variables in the optimization problem
double* tmp = new double[variable_size];
for(int i=0;i<variable_size;i++){
  //initial_solution is a vector stores the initial solution of the variables and 
  //the initial solution satisfies all constraints.
  tmp[i] = initial_solution[i]; 
}
CoinWarmStartVector<double> warm_start_vector(variable_size,tmp);
if(model.setWarmStart(&warm_start_vector)){
  std::cout<<"warm start success"<<std::endl;
}else
  std::cout<<"warm start fail"<<std::endl;
solver.branchAndBound();
```
The code prints the "warm start fail"  and I want to know whether `setWarmStart` function can be used to set the initial solution and when does it return true? Can you give an example of setting the initial solution? Thanks a lot!

**Comment by jhmgoossens** [comment]
Hi,
To solve ILPs I strongly advise you _not to use OsiClp_, but CbcModel.
Next, to set a good initial solution, I use CbcModel::setBestSolution to give CBC a starting solution.
If you know only a good initial solution object value, then use CbcModel::setObjValue.
In both cases CBC can actually find a better solution during solving.
Does this help?

**Reply by ycpengpeng** [reply]
Thank you very much for your answer! It helps me a lot. 
"I strongly advise you not to use OsiClp, but CbcMode" -> Because I know a little aboud ILPS, so I followed the code from the examples/minimum.cpp. I am confused about how to use a CbcModel and I would appreciate it if you could give me an example.
In addition to this, I wonder whether there is an api to set the stop criterion of the solver. To be specific, my problem is like
``` latex
minimize abs(f(x))
``` 
and I want the solver to stop when `abs(f(x))<1e-6` or when the solver has spent more than 60 seconds. Does Cbc support this kind of stop criterion?  
Thank you very much again!

**Reply by jhmgoossens** [reply]
For examples of how to use Cbc, you can look at [modify.cpp for how to load a problem](https://github.com/coin-or/Cbc/blob/master/examples/modify.cpp), and [inc.cpp for how to solve using CbcMain0 and CbcMain1](https://github.com/coin-or/Cbc/blob/master/examples/inc.cpp) and how to setup event handling. Don't use CbcModel::branchAndBound() to solve (as done in minimum.cpp line 33 or modify.cpp line 105), but use CbcMain0 and CbcMain1 as shown in inc.cpp lines 125+126, because in this way you actually call the full branch and cut functionality of Cbc.

Yes, Cbc does have stopping criteria, or you can also handle this yourself using the event handling.
